src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -265,14 +257,107 @@ public NodeStateTable<T> ToImmutableAndFree()
                     return NodeStateTable<T>.Empty;
                 }
 
-                var hasNonCached = _states.Any(static s => s.Any(static i => i.Item2 != EntryState.Cached));
+                var hasNonCached = _states.Any(static s => !s.IsCached);
                 return new NodeStateTable<T>(_states.ToImmutableAndFree(), isCompacted: !hasNonCached);
-
             }
 
             internal ImmutableArray<T> GetLastEntries()
             {
-                return _states[_states.Count - 1].SelectAsArray(t => t.Item1);
+                Debug.Assert(_states.Count > 0);
+                return _states[_states.Count - 1].GetItems();
+            }
+        }
+
+        private readonly struct TableEntry
+        {
+            private static readonly ImmutableArray<EntryState> s_allAddedEntries = ImmutableArray.Create(EntryState.Added);
+            private static readonly ImmutableArray<EntryState> s_allCachedEntries = ImmutableArray.Create(EntryState.Cached);
+            private static readonly ImmutableArray<EntryState> s_allModifiedEntries = ImmutableArray.Create(EntryState.Modified);
+            private static readonly ImmutableArray<EntryState> s_allRemovedEntries = ImmutableArray.Create(EntryState.Removed);
+
+            private readonly ImmutableArray<T> _items;
+            private readonly T? _item;
+            private readonly ImmutableArray<EntryState> _states;
+
+            public TableEntry(T item, EntryState state)
+                : this(item, default, GetSingleArray(state)) { }
+
+            public TableEntry(ImmutableArray<T> items, EntryState state)
+                : this(default, items, GetSingleArray(state)) { }
+
+            private TableEntry(T? item, ImmutableArray<T> items, ImmutableArray<EntryState> states)
+            {
+                Debug.Assert(item is object || !items.IsDefault);
+                Debug.Assert(!states.IsDefault);
+
+                this._item = item;
+                this._items = items;
+                this._states = states;
+            }
+
+            public bool IsCached => this._states == s_allCachedEntries || this._states.All(s => s == EntryState.Cached);
+
+            public bool IsRemoved => this._states == s_allRemovedEntries || this._states.All(s => s == EntryState.Removed);
+
+            public int Count => IsSingle ? 1 : _items.Length;
+
+            public T GetItem(int index)
+            {
+                Debug.Assert(!IsSingle || index == 0);
+                return IsSingle ? _item : _items[index];
+            }
+
+            public EntryState GetState(int index) => _states.Length == 1 ? _states[0] : _states[index];
+
+            public ImmutableArray<T> GetItems() => IsSingle ? ImmutableArray.Create(_item) : _items;
```

Comment:
How often is this called? It defeats the purpose of not using an `ImmutableArray<T>` for storage of `_item` here. Makes me wonder if we should use a different storage mechanism here that we can return an enumerator on that works for both types. The calling code doesn't actually store the resulting value, just enumerates it.

Reply:
Yeah, this is where the abstraction gets a bit leaky and it's actually only used by the syntax node. Let me think we can come up with a better way to handle that case and then we can drop this.

Reply:
Ok, I ended up adding an overload to `TryUsedCachedEntries()` that returns the entries we used. That special cases it so we don't need to get the items during a normal cache lookup. I renamed `GetItems()` to `ToImmutableArray()` to make it clear that this actually performing work as part of the call.

---
In reply to: [638295908](https://github.com/dotnet/roslyn/pull/53568#discussion_r638295908) [](ancestors = 638295908)